### PR TITLE
3137 new token pair endpoint

### PR DESCRIPTION
**Pull request by shreyamalviya**
# What does this PR do?

Fixes a part of #3137 

## PR Checklist
* [x] Have you added an explanation of what your changes do and why you'd like to include them?
* [x] Is the TravisCI build passing?
* [x] Was the CHANGELOG.md updated to reflect the changes?
* [ ] ~Was the documentation framework updated to reflect the changes?~
* [x] Have you checked that you haven't introduced any duplicate code?

## Testing Checklist

* [x] Added relevant unit tests?
* [x] Do all unit tests pass?
* [ ] ~Do all end-to-end tests pass?~
* [x] Any other testing performed?
    > Tested the endpoint manually
* [ ] ~If applicable, add screenshots or log transcripts of the feature working~

### CODE REVIEW

File: monkey/monkey_island/cc/services/authentication_service/flask_resources/__init__.py [hunk 1]
```diff
@@ -5,3 +5,4 @@
 from .register_resources import register_resources
 from .agent_otp import AgentOTP
 from .agent_otp_login import AgentOTPLogin
+from .token import Token
```
Comment (VakarisZ) [inline monkey/monkey_island/cc/services/authentication_service/flask_resources/__init__.py:8]: Wait, what's the purpose of these imports if all resources are registered through `register_resources`?

File: monkey/monkey_island/cc/services/authentication_service/token/token_validator.py [hunk 2]
```diff
@@ -16,3 +16,10 @@ def validate_token(self, token: Token):
         :raises SignatureExpired: If the token has expired
         """
         self._token_serializer.loads(token, max_age=self._token_expiration)
+
+    def get_token_payload(self, token: Token) -> str:
+        """
+        Returns the payload of a token
+        :param token: Token to get the payload of
+        """
+        return str(self._token_serializer.loads(token, max_age=self._token_expiration))
```
Comment (VakarisZ) [inline monkey/monkey_island/cc/services/authentication_service/token/token_validator.py:25]: Typically it's not the responsibility of the validator to decode something, but since we can't have `Manager` and `Service` (Mike doesn't allow us 😞) I don't know what's a better name for this.

File: monkey/tests/unit_tests/monkey_island/cc/services/authentication_service/token/test_token_validator.py [hunk 3]
```diff
@@ -50,3 +50,44 @@ def test_validate_refresh_token__invalid(freezer):
...
+
+def test_get_token_payload__invalid_token():
+    token_expiration = 1 * 60  # 1 minute
+    invalid_token = "invalid_token"
+
+    app, _ = build_app()
+    token_validator = TokenValidator(app.security, token_expiration)
+
+    with pytest.raises(BadData):
+        token_validator.get_token_payload(invalid_token)
```
Comment (VakarisZ) [inline monkey/tests/unit_tests/monkey_island/cc/services/authentication_service/token/test_token_validator.py:93]: I'm not sure we need these tests. Validating the token is done by `validate_token`, I don't think we need to assert this behavior for decoding the payload. Alternatively, we could just have one method to decode the payload and use that for validation as well, but these are different tasks and might change in the future

File: monkey/tests/unit_tests/monkey_island/cc/services/authentication_service/test_authentication_service.py [hunk 4]
```diff
@@ -116,6 +116,46 @@ def test_handle_sucessful_login(
...
+    authentication_facade: AuthenticationFacade,
+):
+    user = User(username=USERNAME, password=PASSWORD, fs_uniquifier="a")
+    refresh_token = authentication_facade.generate_refresh_token(user)
+
+    with pytest.raises(Exception):
+        authentication_facade.generate_new_token_pair(refresh_token)
+
+
+def test_generate_new_token_pair__fails_if_refresh_token_expired(
```
Comment (VakarisZ) [inline monkey/tests/unit_tests/monkey_island/cc/services/authentication_service/test_authentication_service.py:146]:
```suggestion
def test_generate_new_token_pair__fails_if_token_invalid(
```

File: monkey/tests/unit_tests/monkey_island/cc/services/authentication_service/token/test_token_validator.py (hunk 3 above)
Comment (shreyamalviya) [reply monkey/tests/unit_tests/monkey_island/cc/services/authentication_service/token/test_token_validator.py:93]: I think it's worth testing that an error is raised if a bad token is passed to `get_token_payload()`.

**Review by VakarisZ — APPROVED** [review]

**Review by ilija-lazoroski — APPROVED** [review]

File: monkey/tests/unit_tests/monkey_island/cc/services/authentication_service/test_authentication_service.py [hunk 5]
```diff
@@ -72,18 +72,6 @@ def authentication_facade(
...
-def test_needs_registration__true(authentication_facade: AuthenticationFacade):
-    assert authentication_facade.needs_registration()
-
-
-def test_needs_registration__false(
-    monkeypatch,
-    authentication_facade: AuthenticationFacade,
-):
-    User(username=USERNAME, password=PASSWORD).save()
-    assert not authentication_facade.needs_registration()
```
[the remark this answers is not in the document]
Comment (mssalvatore) [reply monkey/tests/unit_tests/monkey_island/cc/services/authentication_service/test_authentication_service.py:84]: This reduces test coverage on a security-critical component of the ecosystem. I'm adding these back.
image

**Review by mssalvatore — CHANGES_REQUESTED** [review]
I added the "needs registration" tests back in and they fail. These tests MUST be here and they MUST pass.

File: monkey/monkey_island/cc/services/authentication_service/token/token_validator.py (hunk 2 above)
Comment (mssalvatore) [reply monkey/monkey_island/cc/services/authentication_service/token/token_validator.py:25]: How about `TokenParser`. `TokenParser.parse()` can return a data structure containing the components of the token. It can raise validation errors if the token is invalid.

File: monkey/monkey_island/cc/services/authentication_service/flask_resources/token.py [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+
+logger = logging.getLogger(__name__)
+
+
+class Token(AbstractResource):
+    """
+    A resource for refreshing tokens
+    """
+
+    urls = ["/api/token"]
```
Comment (mssalvatore) [inline monkey/monkey_island/cc/services/authentication_service/flask_resources/token.py:19]: This name is pretty meaningless. It's also violates RESTful principles. If I POST to `/api/token`, I should be passing it some data to use to create a "token" resource. 

This is effectively a login/authentication call. It is not RESTful pretty much by definition; it's RPC. If it's an RPC call, it should have an RPC name.

Suggest: `POST /api/refresh-authentication-token`

File: monkey/tests/unit_tests/monkey_island/cc/services/authentication_service/test_authentication_service.py [hunk 7]
```diff
@@ -116,6 +116,47 @@ def test_handle_sucessful_login(
...
+    refresh_token = "original_token"
+    user = User(username=USERNAME, password=PASSWORD, fs_uniquifier="a")
+    user.save()
+    mock_token_generator.generate_token.return_value = "new_token"
+    mock_token_validator.get_token_payload.return_value = "a"
+    access_token, new_refresh_token = authentication_facade.generate_new_token_pair(refresh_token)
+
+    assert new_refresh_token != refresh_token
+    assert access_token != refresh_token
+    assert access_token != new_refresh_token
```
Comment (mssalvatore) [inline monkey/tests/unit_tests/monkey_island/cc/services/authentication_service/test_authentication_service.py:133]: Missing a check or test that verifies the newly returned authentication token is different from the user's original authentication token.

File: monkey/tests/unit_tests/monkey_island/cc/services/authentication_service/token/test_token_validator.py (hunk 3 above)
Comment (mssalvatore) [reply monkey/tests/unit_tests/monkey_island/cc/services/authentication_service/token/test_token_validator.py:93]: +1, worth the test. But this will probably change anyway if we [change it to `TokenParser`.](https://github.com/guardicore/monkey/pull/3181#discussion_r1154679927)

File: monkey/tests/unit_tests/monkey_island/cc/services/authentication_service/test_authentication_service.py [hunk 8]
```diff
@@ -62,25 +71,28 @@ def authentication_facade(
...
+    authentication_facade._datastore.find_user.return_value = False
     assert authentication_facade.needs_registration()
 
 
 def test_needs_registration__false(
     monkeypatch,
     authentication_facade: AuthenticationFacade,
 ):
-    User(username=USERNAME, password=PASSWORD).save()
+    authentication_facade._datastore.find_user.return_value = True
```
Comment (mssalvatore) [inline monkey/tests/unit_tests/monkey_island/cc/services/authentication_service/test_authentication_service.py:95]: Why is this change necessary? Doesn't `UserDatastore` rely on the `User` model? If we add or remove `Users` directly using the `User` model, shouldn't `UserDatastore` return the expected values?

I don't necessarily object to these changes, but I find it concerning that they need to be made.

**Review by mssalvatore — APPROVED** [review]

File: monkey/tests/unit_tests/monkey_island/cc/services/authentication_service/test_authentication_service.py (hunk 8 above)
Comment (shreyamalviya) [reply monkey/tests/unit_tests/monkey_island/cc/services/authentication_service/test_authentication_service.py:88]: We're now using `UserDatastore` instead of the `User` model in the `needs_registration()` method. See 59d5c0707c441adf0f471b481a725a2d87998e3a. The datastore is mocked in these tests so the return value needs to be specified.
